pootle/apps/pootle_store/unit/timeline.py

```diff
@@ -0,0 +1,218 @@
+            if self.full_name.strip()
+            else self.username)
+
+    @property
+    def user_url(self):
+        return reverse(
+            'pootle-user-profile',
+            args=[self.username])
+
+
+class TimelineEntry(object):
+
+    def __init__(self, item, new_value=None):
+        self.item = item
+
+    @property
+    def entry_dict(self):
+        return {
+            'field': self.item['field'],
+            'field_name': SubmissionFields.NAMES_MAP.get(self.item['field'], None),
+            'type': self.item['type']}
+
+    def base_entry(self):
+        entry = self.entry_dict
+        entry['new_value'] = to_python(self.item['new_value'])
+        return entry
+
+    def qc_entry(self):
+        entry = self.entry_dict
+        check_name = self.item['quality_check__name']
+        check_url = (
+            u''.join(
+                [reverse('pootle-checks-descriptions'),
+                 '#', check_name]))
+        entry.update(
+            {'check_name': check_name,
+             'check_display_name': check_names[check_name],
+             'checks_url': check_url})
+        return entry
+
+    def state_change_entry(self):
+        entry = self.entry_dict
+        entry['old_value'] = STATES_MAP[int(to_python(self.item['old_value']))]
+        entry['new_value'] = STATES_MAP[int(to_python(self.item['new_value']))]
+        return entry
+
+    def suggestion_entry(self):
+        entry = self.entry_dict
+        suggestion_description = mark_safe(
+            SuggestionEvent(
+                self.item["type"],
+                self.item["suggestion__user__username"],
+                self.item["suggestion__user__full_name"]).description)
+        entry.update(
+            {'suggestion_text': self.item['suggestion__target_f'],
+             'suggestion_description': suggestion_description})
+        return entry
+
+    @property
+    def entry(self):
+        if self.item['field'] == SubmissionFields.STATE:
+            return self.state_change_entry()
+        elif self.item['suggestion_id']:
+            return self.suggestion_entry()
+        elif self.item['quality_check__name']:
+            return self.qc_entry()
+        return self.base_entry()
+
+
+class Timeline(object):
+
+    entry_class = TimelineEntry
+    timeline_fields = [
+        "type", "old_value", "new_value", "submitter_id", "creation_time",
+        "field", "quality_check_id",
+        "submitter__username", "submitter__email", "submitter__full_name",
+        "suggestion_id", "suggestion__target_f", "suggestion__user__full_name",
+        "suggestion__user__username", "quality_check__name"]
+
+    def __init__(self, object):
+        self.object = object
+
+    @property
+    def grouped_entries(self):
+        grouped_entries = self.get_grouped_timeline_entries()
+        grouped_entries = self.add_creation_entry(grouped_entries)
+        grouped_entries.reverse()
+        return grouped_entries
+
+    @property
+    def timeline(self):
```

Comment:
Being `Timeline` the name of the class, it's probably redundant to have `timeline*` in contained attributes and methods. For this specific case, what about naming the method `events`?